src/cluster_slot.c

```diff
@@ -0,0 +1,207 @@
+#include "server.h"
+#include "cluster.h"
+#include "cluster_slot.h"
+
+static uint64_t getSingleSlotStat(int slot, int order_by) {
+    serverAssert(order_by != ORDER_BY_INVALID);
+    uint64_t singleSlotStat = 0;
+    if (order_by == ORDER_BY_KEY_COUNT) {
+        singleSlotStat = (*server.db->slots_to_keys).by_slot[slot].count;
+    }
+    return singleSlotStat;
+}
+
+static void addReplySingleSlotStat(client *c, int slot) {
+    addReplyLongLong(c, slot);
+    addReplyMapLen(c, 1);
+    addReplyBulkCString(c, "key_count");
+    addReplyLongLong(c, (*server.db->slots_to_keys).by_slot[slot].count);
+}
+
+static void addReplySortedSlotStats(client *c, sortedSlotStatEntry sorted[], int num_slots_assigned, int limit) {
+    int len = min(limit, num_slots_assigned);
+    addReplyMapLen(c, len);
+
+    for (int i = 0; i < len; i++) {
+        addReplySingleSlotStat(c, sorted[i].slot);
+    }
+}
+
+static int countValidSlots(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlots() {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (server.cluster->slots[slot]) count++;
+    }
+    return count;
+}
+
+static int slotStatEntryAscCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_a.stat - entry_b.stat;
+}
+
+static int slotStatEntryDescCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_b.stat - entry_a.stat;
+}
+
+static int checkSlotAssignment(client *c, unsigned char *slots, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (server.cluster->slots[slot]) {
+            if (slots[slot]++ == 1) {
+                addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
+                return C_ERR;
+            }
+        }
+    }
+    return C_OK;
+}
+
+void addReplySlotStats(client *c, unsigned char *slots) {
+    int count = countValidSlots(slots);
+    addReplyMapLen(c, count);
+
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (!slots[slot]) continue;
+        addReplySingleSlotStat(c, slot);
```

Comment:
Is this preferred over something like `if (slots[slot]) addReplySingleSlotStat(c, slot);`?